File: packages/pulsar/src/session-vault.ts
import type { Account, SessionLoginCode, SessionResult } from './address-book.ts';
import { AddressBookRequestError, type AddressBookClient } from './address-book-client.ts';
import { sessionKeyMessage, sessionRefreshMessage } from './signing.ts';

/* What is kept between starts. The access token is not: it lives a quarter of an hour and a refresh makes a new one. */
export interface StoredSession {
    refreshToken: string;
    // When the session ends whatever happens, counted from sign-in.
    expiresAt: number;
    account: Account;
}

/* Where the refresh token lives: the shell's encrypted file, a page's IndexedDB, a phone's keychain, memory in a test. */
export interface SessionStore {
    read(): Promise<StoredSession | null>;
    write(session: StoredSession | null): Promise<void>;
}

/*
 * The key a session is bound to. The private half never leaves the side that holds it: a
 * non-extractable `CryptoKey` in a page, a key encrypted with the OS keychain in the desktop shell.
 */
export interface SessionSigner {
    /* The raw public key in base64url, the shape `PublicKeySchema` checks. */
    publicKey: string;
    sign(message: string): Promise<string>;
}

/* A session as the side that uses it sees it: everything but the refresh token. */
export interface SessionView {
    accessToken: string;
    accessExpiresAt: number;
    expiresAt: number;
    account: Account;
}

export interface SessionVaultOptions {
    client: AddressBookClient;
    store: SessionStore;
    /* Null when this side has no key any more, which leaves a kept session nothing to refresh with. */
    signer: () => Promise<SessionSigner | null>;
    now?: () => number;
}

const viewOf = (result: SessionResult): SessionView => ({
    accessToken: result.accessToken,
    accessExpiresAt: result.accessExpiresAt,
    expiresAt: result.expiresAt,
    account: result.account
});

/*
 * The one holder of an address book refresh token. It trades a login code for a session bound to its
 * signer's key, hands out access tokens and never the refresh token, and rotates it on every refresh.
 * Refreshes are one at a time: the address book ends a session when a spent refresh token comes back,
 * so two refreshes racing with the same token would sign the person out.
 */
export class SessionVault {
    private readonly client: AddressBookClient;
    private readonly store: SessionStore;
    private readonly signer: () => Promise<SessionSigner | null>;
    private readonly now: () => number;
    private current: SessionView | null = null;
    private refreshing: Promise<SessionView | null> | null = null;

    constructor(options: SessionVaultOptions) {
        this.client = options.client;
        this.store = options.store;
        this.signer = options.signer;
        this.now = options.now ?? Date.now;
    }

    async exchange(login: SessionLoginCode): Promise<SessionView> {
        const signer = await this.signer();
        if (!signer) {
            throw new Error('This device has no key to bind a session to');
        }
        const result = await this.client.exchange({
            ...login,
            sessionKey: signer.publicKey,
            sessionKeySignature: await signer.sign(sessionKeyMessage(login.code, signer.publicKey))
        });
        await this.keep(result);
        return this.current!;
    }

    /* Who is signed in, from what was kept, without asking the address book. */
    async restore(): Promise<{ account: Account; expiresAt: number } | null> {
        const stored = await this.store.read();
        if (!stored) {
            return null;
        }
        if (stored.expiresAt <= this.now()) {
            await this.store.write(null);
            return null;
        }
        return { account: stored.account, expiresAt: stored.expiresAt };
    }

    /*
     * A fresh access token. Null when there is no session any more (never signed in, ended, no key to
     * sign with, or the address book refused the token), which clears what was kept; a network failure
     * throws and keeps the session, since the token is still good once the address book answers again.
     */
    refresh(): Promise<SessionView | null> {
        this.refreshing ??= this.rotate().finally(() => {
            this.refreshing = null;
        });
        return this.refreshing;
    }

    /* Ends the session at the address book when it can, and forgets it here whatever the answer. */
    async signOut(): Promise<void> {
        const token = this.current?.accessToken ?? (await this.refresh().catch(() => null))?.accessToken ?? null;
        this.current = null;
        await this.store.write(null);
        if (token !== null) {
            await this.client.endSession(token).catch(() => undefined);
        }
    }

    private async rotate(): Promise<SessionView | null> {
        const stored = await this.store.read();
        const signer = stored && stored.expiresAt > this.now() ? await this.signer() : null;
        if (!stored || !signer) {
            this.current = null;
            if (stored) {
                await this.store.write(null);
            }
            return null;
        }
        const issuedAt = this.now();
        let result: SessionResult;
        try {
            result = await this.client.refresh({
                refreshToken: stored.refreshToken,
                issuedAt,
                signature: await signer.sign(sessionRefreshMessage(stored.refreshToken, issuedAt))
            });
        } catch (e) {
            if (e instanceof AddressBookRequestError && (e.code === 'unauthorized' || e.code === 'bad-request' || e.code === 'bad-signature')) {
                this.current = null;
                await this.store.write(null);
                return null;
            }
            throw e;
        }
        await this.keep(result);
        return this.current;
    }

    private async keep(result: SessionResult): Promise<void> {
        await this.store.write({ refreshToken: result.refreshToken, expiresAt: result.expiresAt, account: result.account });
        this.current = viewOf(result);
    }
}

File: apps/ios/Packages/RuimtePulsar/Sources/RuimtePulsar/SessionVault.swift
import Foundation

public struct StoredSession: Codable, Sendable {
    public let refreshToken: String
    public let expiresAt: Int64
    public let account: Account

    public init(refreshToken: String, expiresAt: Int64, account: Account) {
        self.refreshToken = refreshToken
        self.expiresAt = expiresAt
        self.account = account
    }
}

public protocol SessionStore: Sendable {
    func read() throws -> StoredSession?
    func write(_ session: StoredSession?) throws
}

public struct SessionView: Sendable {
    public let accessToken: String
    public let accessExpiresAt: Int64
    public let expiresAt: Int64
    public let account: Account
}

public struct RestoredSession: Sendable {
    public let account: Account
    public let expiresAt: Int64
}

public actor SessionVault {
    private let client: any SessionAPI
    private let store: any SessionStore
    private let signer: @Sendable () async throws -> (any SessionSigner)?
    private let now: @Sendable () -> Int64
    private var current: SessionView?
    private var currentRevision: Int?
    private var refreshing: Task<SessionView?, Error>?
    private var revision = 0

    public init(
        client: any SessionAPI, store: any SessionStore,
        signer: @escaping @Sendable () async throws -> (any SessionSigner)?,
        now: @escaping @Sendable () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.client = client
        self.store = store
        self.signer = signer
        self.now = now
    }

    public func exchange(_ login: SessionLoginCode) async throws -> SessionView {
        try Task.checkCancellation()
        revision += 1
        let operation = revision
        refreshing?.cancel()
        refreshing = nil
        guard let key = try await signer() else {
            throw AddressBookRequestError(
                code: "unauthorized", status: 0, message: "This device has no key to bind a session to.")
        }
        try Task.checkCancellation()
        guard operation == revision else { throw CancellationError() }
        let signature = try key.sign(SigningBytes.sessionKey(code: login.code, publicKey: key.publicKey))
        let result = try await client.exchange(
            SessionExchangePayload(
                code: login.code, codeVerifier: login.codeVerifier, redirectUri: login.redirectUri,
                label: login.label, sessionKey: key.publicKey, sessionKeySignature: signature
            ))
        try Task.checkCancellation()
        guard operation == revision else {
            throw CancellationError()
        }
        return try keep(result)
    }

    public func restore() throws -> RestoredSession? {
        guard let stored = try store.read() else {
            return nil
        }
        guard stored.expiresAt > now() else {
            try store.write(nil)
            return nil
        }
        return RestoredSession(account: stored.account, expiresAt: stored.expiresAt)
    }

    public func accessToken() async throws -> String? {
        if let current, current.accessExpiresAt > now() + 30_000 {
            return current.accessToken
        }
        return try await refresh()?.accessToken
    }

    public func refresh() async throws -> SessionView? {
        try await refreshOperation().value
    }

    func refreshOperation() -> Task<SessionView?, Error> {
        if let refreshing {
            return refreshing
        }
        let operation = revision
        let task = Task {
            defer {
                if operation == revision {
                    refreshing = nil
                }
            }
            return try await rotate(operation: operation)
        }
        refreshing = task
        return task
    }

    public func signOut() async throws {
        let token: String?
        if let current {
            token = current.accessToken
        } else {
            token = try? await refresh()?.accessToken
        }
        revision += 1
        refreshing?.cancel()
        refreshing = nil
        current = nil
        currentRevision = nil
        try store.write(nil)
        if let token {
            try? await client.endSession(accessToken: token)
        }
    }

    public func discard(accessToken: String) async {
        // A canceled login must never erase a newer login that finished during cleanup.
        if current?.accessToken == accessToken {
            if currentRevision == revision { revision += 1 }
            refreshing?.cancel()
            refreshing = nil
            current = nil
            currentRevision = nil
            try? store.write(nil)
        }
        try? await client.endSession(accessToken: accessToken)
    }

    private func rotate(operation: Int) async throws -> SessionView? {
        let stored = try store.read()
        let key = if let stored, stored.expiresAt > now() { try await signer() } else { nil as (any SessionSigner)? }
        guard operation == revision else {
            throw CancellationError()
        }
        guard let stored, let key else {
            current = nil
            if stored != nil {
                try store.write(nil)
            }
            return nil
        }
        let issuedAt = now()
        let signature = try key.sign(SigningBytes.sessionRefresh(token: stored.refreshToken, issuedAt: issuedAt))
        let result: SessionResult
        do {
            result = try await client.refresh(
                SessionRefreshPayload(refreshToken: stored.refreshToken, issuedAt: issuedAt, signature: signature))
        } catch let error as AddressBookRequestError
            where ["unauthorized", "bad-request", "bad-signature"].contains(error.code)
        {
            guard operation == revision else {
                throw CancellationError()
            }
            current = nil
            try store.write(nil)
            return nil
        }
        try Task.checkCancellation()
        // A late response must not restore a session after sign-out or another login.
        guard operation == revision else {
            throw CancellationError()
        }
        return try keep(result)
    }

    private func keep(_ result: SessionResult) throws -> SessionView {
        try store.write(
            StoredSession(refreshToken: result.refreshToken, expiresAt: result.expiresAt, account: result.account))
        let view = SessionView(
            accessToken: result.accessToken, accessExpiresAt: result.accessExpiresAt, expiresAt: result.expiresAt,
            account: result.account)
        current = view
        currentRevision = revision
        return view
    }
}
